ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/transaction/UtxoLedgerTransaction.kt

```diff
@@ -0,0 +1,189 @@
+ * @property outputStateAndRefs The output states and state references to be created by the current [UtxoLedgerTransaction].
+ * @property outputTransactionStates The output transaction states to be created by the current [UtxoLedgerTransaction].
+ * @property outputContractStates The output contract states to be created by the current [UtxoLedgerTransaction].
+ */
+@DoNotImplement
+@Suppress("TooManyFunctions")
+interface UtxoLedgerTransaction {
+
+    val timeWindow: TimeWindow?
+    val attachments: List<Attachment>
+    val commands: List<Command>
+    val signatories: Set<PublicKey>
+
+    val inputStateAndRefs: List<StateAndRef<*>>
+    val inputTransactionStates: List<TransactionState<*>> get() = inputStateAndRefs.map { it.state }
+    val inputContractStates: List<ContractState> get() = inputStateAndRefs.map { it.state.contractState }
+
+    val referenceInputStateAndRefs: List<StateAndRef<*>>
+    val referenceInputTransactionStates: List<TransactionState<*>> get() = referenceInputStateAndRefs.map { it.state }
+    val referenceInputContractStates: List<ContractState> get() = referenceInputTransactionStates.map { it.contractState }
+
+    val outputStateAndRefs: List<StateAndRef<*>>
+    val outputTransactionStates: List<TransactionState<*>> get() = outputStateAndRefs.map { it.state }
+    val outputContractStates: List<ContractState> get() = outputTransactionStates.map { it.contractState }
+
+    /**
+     * Verifies the current [UtxoLedgerTransaction].
+     *
+     * @throws TransactionVerificationException if the current [UtxoLedgerTransaction] fails to verify correctly.
+     */
+    fun verify()
+
+    /**
+     * Obtains the ledger transaction [Attachment] with the specified id.
+     *
+     * @param id The id of the ledger transaction [Attachment] to obtain.
+     * @return Returns the ledger transaction [Attachment] with the specified id.
+     * @throws IllegalArgumentException if the ledger transaction [Attachment] with the specified id cannot be found.
+     */
+    fun getAttachment(id: SecureHash): Attachment
+
+    /**
+     * Obtains all ledger transaction [Command] instances that match the specified type.
+     *
+     * @param T The underlying type of the [Command].
+     * @param type The type of the [Command].
+     * @return Returns all ledger transaction [Command] instances that match the specified type.
+     */
+    fun <T : Command> getCommands(type: Class<T>): List<T>
+
+    /**
+     * Obtains all ledger transaction [StateAndRef] inputs that match the specified [ContractState] type.
+     *
+     * @param T The underlying type of the [ContractState].
+     * @param type The type of the [ContractState].
+     * @return Returns all ledger transaction [StateAndRef] inputs that match the specified [ContractState] type.
+     */
+    fun <T : ContractState> getInputStateAndRefs(type: Class<T>): List<StateAndRef<T>>
```

Comment:
Do we need all these methods that resolve e.g. a StateRef to a StateAndRef on the LedgerTx?

Reply:
AFAIK no such methods on the original `LedgerTransaction` and no equivalent here. We only have the bare minimum APIs we need.